docs/annotated_mnist.md

```diff
@@ -1,157 +1,317 @@
-  return optimizer
+  metrics = compute_metrics(logits, batch['label'])
+  return optimizer, metrics
 ```
 
-Making model predictions is as simple as calling `model(input)`:
+2. Create a [`jit`](https://jax.readthedocs.io/en/latest/jax.html#jax.jit)-compiled function that evaluates the model on the test set using [`flax.linen.apply`](https://flax.readthedocs.io/en/latest/flax.linen.html#flax.linen.Module.apply):
 
-```py
+
+```python
 @jax.jit
-def eval(model, eval_ds):
-  logits = model(eval_ds['image'] / 255.0)
-  return compute_metrics(logits, eval_ds['label'])
+def eval_step(params, batch):
+  logits = CNN().apply({'params': params}, batch['image'])
+  return compute_metrics(logits, batch['label'])
 ```
 
-## Main train loop
+3. Define a training function for one epoch that:
+
+  - Shuffles the training data before each epoch using [`jax.random.permutation`](https://jax.readthedocs.io/en/latest/_autosummary/jax.random.permutation.html) that takes a PRNGKey as a parameter (discussed in more detail later in this tutorial and in [JAX - the sharp bits](https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html#JAX-PRNG)).
+  - Runs an optimization step for each batch.
+  - Retrieves the training metrics from the device with `jax.device_get` and computes their mean across each batch in an epoch.
+  - Returns the optimizer with updated parameters and the training loss and accuracy metrics.
+
+
+```python
+def train_epoch(optimizer, train_ds, batch_size, epoch, rng):
+  train_ds_size = len(train_ds['image'])
+  steps_per_epoch = train_ds_size // batch_size
+
+  perms = jax.random.permutation(rng, len(train_ds['image']))
+  perms = perms[:steps_per_epoch * batch_size]  # Skip an incomplete batch
+  perms = perms.reshape((steps_per_epoch, batch_size))
+
+  batch_metrics = []
+
+  for perm in perms:
+    batch = {k: v[perm, ...] for k, v in train_ds.items()}
+    optimizer, metrics = train_step(optimizer, batch)
+    batch_metrics.append(metrics)
+
+  training_batch_metrics = jax.device_get(batch_metrics)
+  training_epoch_metrics = {
+      k: np.mean([metrics[k] for metrics in training_batch_metrics])
+      for k in training_batch_metrics[0]}
+
+  print('Training - epoch: %d, loss: %.4f, accuracy: %.2f' % (epoch, training_epoch_metrics['loss'], training_epoch_metrics['accuracy'] * 100))
+
+  return optimizer, training_epoch_metrics
+```
+
+4. Create a model evaluation function that:
+
+  - Evalues the model on the test set.
+  - Retrieves the evaluation metrics from the device with `jax.device_get`.
+  - Copies the metrics [data stored](https://flax.readthedocs.io/en/latest/design_notes/linen_design_principles.html#how-are-parameters-represented-and-how-do-we-handle-general-differentiable-algorithms-that-update-stateful-variables) in a JAX [pytree](https://jax.readthedocs.io/en/latest/pytrees.html#pytrees-and-jax-functions).
+  - Returns the test loss and accuracy.
 
-```py
-def train():
+
+```python
+def eval_model(model, test_ds):
```

Comment:
The argument should be `params`, not `model`.

Reply:
The code is from our `linen_examples`:

Source: https://github.com/google/flax/blob/9fdf59f05702c746ea9a6be50cd3a81f38e2fa0b/linen_examples/mnist/train.py#L131

```py
def eval_model(model, test_ds):
  metrics = eval_step(model, test_ds)
  metrics = jax.device_get(metrics)
  summary = jax.tree_map(lambda x: x.item(), metrics)
  return summary['loss'], summary['accuracy']
```

Let me know if I should still change this